### PR TITLE
FEAT Adding experimental feature : Triton mm int8xint2 

### PR DESCRIPTION
**Pull request by MekkCyber**
### Summary
Introducing matrix multiplication int8xint2 in Triton as an experimental feature. This approach involves performing matmul with on-the-fly unpacking, utilizing cached tiling techniques. Currently, it leverages tl.dot with int8 values, which is the most optimized method available at this time. However, with future hardware advancements, this could become significantly more efficient, particularly when using ternary weights, potentially eliminating the need for multiplication altogether.

### CODE REVIEW

File: src/liger_kernel/ops/experimental/mm_int8int2.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        triton.Config({'BLOCK_SIZE_M': 32, 'BLOCK_SIZE_N': 32, 'BLOCK_SIZE_K': 32, 'GROUP_SIZE_M': 4}, num_stages=4,
+                      num_warps=4)
+    ]
+
+@triton.autotune(
+    configs=get_cuda_autotune_config(),
+    key=['M', 'N', 'K'],
+)
+@triton.jit
+def matmul_kernel(
```
Comment (ByronHsu) [inline src/liger_kernel/ops/experimental/mm_int8int2.py:93]: add a reference and more comments. it is a bit hard to understand the current code

File: src/liger_kernel/ops/experimental/mm_int8int2.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        a, b, c,
+        M, N, K,
+        a.stride(0), a.stride(1),
+        b.stride(0), b.stride(1),
+        c.stride(0), c.stride(1),
+    )
+    return c
+
+
+def test_kernel(size=2048) :
```
Comment (ByronHsu) [inline src/liger_kernel/ops/experimental/mm_int8int2.py:169]: move this to tests/

**Review by ByronHsu — COMMENTED** [review]
awesome work!

File: src/liger_kernel/ops/experimental/mm_int8int2.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    it = min(values_per_item, (original_shape[0] // row_dim) + 1)
+    for i in range(it):
+        start = i * row_dim
+        end = min(start + row_dim, original_shape[0])
+        packed[: (end - start)] |= lshift(unpacked[start:end], bits * i)
+
+    return packed
+
+
+def get_cuda_autotune_config():
```
Comment (ByronHsu) [inline src/liger_kernel/ops/experimental/mm_int8int2.py:54]: should not be cuda?

**Comment by ByronHsu** [comment]
good to merge if we can add more comments / ref to explain

File: test/transformers/test_mm_int8int2.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    u = torch.randint(0, 255, (out_features, size), device=device, dtype=torch.uint8)
+
+    # Calculate dimensions
+    B, M, N = ht.size()
+
+    # Compute triton output
+    triton_output = matmul(ht.view(B * M, N), u.T.contiguous()).view(B, M, -1)
+
+    # Validate packing and unpacking of weights
+    assert (pack_weights(unpack_weights(u.T), 2) == u.T).all(), "Packed weights do not match original weights."
```
Comment (ByronHsu) [inline test/transformers/test_mm_int8int2.py:64]: imo we can separate the correctness of pack + unpack to another testing func

File: src/liger_kernel/ops/experimental/mm_int8int2.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+def matmul_kernel(
+        a_ptr, b_ptr, c_ptr,
+        M, N, K,
+        stride_am, stride_ak,
+        stride_bk, stride_bn, 
+        stride_cm, stride_cn,
+        BLOCK_SIZE_M: tl.constexpr, BLOCK_SIZE_N: tl.constexpr, BLOCK_SIZE_K: tl.constexpr,  
+        GROUP_SIZE_M: tl.constexpr,
+):
+    # Only triggered when TRITON_DEBUG is set to 1 => ex : TRITON_DEBUG=1 python scritp.py
```
Comment (ByronHsu) [inline src/liger_kernel/ops/experimental/mm_int8int2.py:102]: what is this?

File: test/transformers/test_mm_int8int2.py (hunk 4 above)
Comment (MekkCyber) [reply test/transformers/test_mm_int8int2.py:64]: Okay i will

File: src/liger_kernel/ops/experimental/mm_int8int2.py (hunk 5 above)
Comment (MekkCyber) [reply src/liger_kernel/ops/experimental/mm_int8int2.py:102]: it's a device_assert, it only works when running with TRITON_DEBUG not set to 0, and it ensure that the K is a multiple of BLOCK_SIZE * 4, which is the case of weight matrices, for alignment purposes. In the future we can find a way to make it more general

File: src/liger_kernel/ops/experimental/mm_int8int2.py (hunk 3 above)
Comment (MekkCyber) [reply src/liger_kernel/ops/experimental/mm_int8int2.py:54]: Sorry, I didn't understand the comment

File: src/liger_kernel/ops/experimental/mm_int8int2.py (hunk 1 above)
Comment (MekkCyber) [reply src/liger_kernel/ops/experimental/mm_int8int2.py:93]: Yes I will add some comments to explain the process when i have some time

File: src/liger_kernel/ops/experimental/mm_int8int2.py (hunk 3 above)
Comment (ByronHsu) [reply src/liger_kernel/ops/experimental/mm_int8int2.py:54]: i mean triton and cuda are different things. maybe replace cuda with triton

File: src/liger_kernel/ops/experimental/mm_int8int2.py (hunk 5 above)
Comment (ByronHsu) [reply src/liger_kernel/ops/experimental/mm_int8int2.py:102]: Does that mean it doesn't error out even if the alignment is incorrect if TRITON_DEBUG not enabled? Wondering if we can use https://triton-lang.org/main/python-api/generated/triton.language.static_assert.html#triton.language.static_assert

Comment (ByronHsu) [reply src/liger_kernel/ops/experimental/mm_int8int2.py:102]: or can we just uplift the assertion before the kernel launch (line 158)

Comment (MekkCyber) [reply src/liger_kernel/ops/experimental/mm_int8int2.py:102]: Now it's working with static_assert, I just had to specify that K is tl.constexpr

File: src/liger_kernel/ops/experimental/mm_int8int2.py (hunk 1 above)
Comment (ByronHsu) [reply src/liger_kernel/ops/experimental/mm_int8int2.py:93]: claude / chatgpt can be helpful

**Review by qingquansong — COMMENTED** [review]
For my own learning purpose, any specific reason we choose bit2 here? I think 1-bit has some cuda support bit counting or matching operations so could be faster if implemented in the context of CUDA, 4bits is safer when doing weight quantization, and is bit 2 a trade-off of accuracy and speed compared bot 4bits and 1bits?

**Review by ByronHsu — DISMISSED** [review]
LGTM

**Comment by ByronHsu** [comment]
cc @MekkCyber please fix the style & test and we are good to merge

**Comment by ByronHsu** [comment]
@MekkCyber ci still failing

**Comment by MekkCyber** [comment]
weird, all tests pass locally, will look into that

**Comment by ByronHsu** [comment]
@MekkCyber can you follow up on this? we can merge it for the next release

**Review by ByronHsu — APPROVED** [review]

**Pull request merged**